pkg/apis/eventing/v1alpha1/channel_types.go

```diff
@@ -0,0 +1,172 @@
+}
+
+// Check that Channel can be validated, can be defaulted, and has immutable fields.
+var _ apis.Validatable = (*Channel)(nil)
+var _ apis.Defaultable = (*Channel)(nil)
+var _ apis.Immutable = (*Channel)(nil)
+var _ runtime.Object = (*Channel)(nil)
+var _ webhook.GenericCRD = (*Channel)(nil)
+
+// ChannelSpec specifies the Provisioner backing a channel and the configuration
+// arguments for a Channel.
+type ChannelSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// Provisioner defines the name of the Provisioner backing this channel.
+	// TODO: +optional If missing, a default Provisioner may be selected for the Channel.
+	Provisioner *ProvisionerReference `json:"provisioner,omitempty"`
+
+	// Arguments defines the arguments to pass to the Provisioner which provisions
+	// this Channel.
+	// +optional
+	Arguments *runtime.RawExtension `json:"arguments,omitempty"`
+}
+
+// ChannelStatus represents the current state of a Channel.
+type ChannelStatus struct {
+	// ObservedGeneration is the most recent generation observed for this Channel.
+	// It corresponds to the Channel's generation, which is updated on mutation by
+	// the API Server.
+	//TODO The above comment is only true once
+	// https://github.com/kubernetes/kubernetes/issues/58778 is fixed.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// A reference to the k8s Service backing this channel.
+	// +optional
+	Service *corev1.LocalObjectReference `json:"service,omitempty"`
+
+	// A reference to the istio VirtualService backing this channel.
+	// +optional
+	VirtualService *corev1.LocalObjectReference `json:"virtualService,omitempty"`
+
+	// DomainInternal holds the top-level domain that will distribute traffic
+	// over the provided targets from inside the cluster. It generally has the
+	// form {channel}.{namespace}.svc.cluster.local
+	//TODO move this to a struct that can be embedded similar to ObjectMeta and
+	// TypeMeta.
+	// +optional
+	DomainInternal string `json:"domainInternal,omitempty"`
+
+	// Represents the latest available observations of a channel's current state.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions []ChannelCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+type ChannelConditionType string
+
+const (
+	// ChannelConditionReady has status True when the Channel is ready to accept traffic.
+	ChannelConditionReady ChannelConditionType = "Ready"
+
+	// ChannelConditionServiceable has status True when the service addressing the
+	// Channel exists.
+	// TODO should this be Sinkable?
+	ChannelConditionServiceable ChannelConditionType = "Serviceable"
```

Comment:
Since a channel may not be fronted by a Service, "Serviceable" is no longer a good name. "Sinkable" (or whatever we end up calling that interface) is good. I'd also consider "Targetable" since that aligns with the .status.domainInternal value being set.

Reply:
I removed this condition since I can't think of a reason for it to be different from Ready.